### ISSUE TITLE
Build A Privacy-Safe Conversation Analytics And Training Dataset

### ISSUE THREAD

**Issue by lobolanja** [lobolanja/BotForge#25]
# Build A Privacy-Safe Conversation Analytics And Training Dataset

## Summary

Create a separate dataset from user conversations so the team can understand how
people interact with the bot and improve future versions without exposing user
identity or private content unnecessarily.

This is not the same as operational message storage. Operational storage is used
to answer users and recover from failures. This dataset is used for product
analytics, quality review, prompt improvement, and possible future model
training.

## Why This Matters

If the product works, users will tell the bot what they need in natural
language. That information is valuable:

- It shows what users are trying to do.
- It reveals confusing flows.
- It identifies missing commands or missing domain knowledge.
- It helps evaluate answer quality.
- It can support future fine-tuning or evaluation datasets.

But free-text conversations can include names, phone numbers, addresses, health
information, financial information, business secrets, and other sensitive
details. We must design this carefully from the start.

## Important Privacy Distinction

Do not assume that removing `telegram_id` is enough.

- Anonymized data means the person is not identifiable anymore.
- Pseudonymized data means direct identifiers were replaced, but the person
  could still be identified using additional information.
- Pseudonymized data is still personal data under GDPR.
- Free-text chat messages are hard to anonymize because users can write personal
  details inside the message body.

For the MVP, name the internal system as a de-identified analytics dataset
unless the team has validated that the output is truly anonymous.

References:

- https://commission.europa.eu/law/law-topic/data-protection/data-protection-explained_en
- https://www.aepd.es/en/prensa-y-comunicacion/blog/anonymisation-and-pseudonymisation
- https://www.aepd.es/prensa-y-comunicacion/notas-de-prensa/guia-y-herramienta-basica-de-anonimizacion

## Product Goal

The team should be able to answer questions like:

- What are users asking the bot?
- Which intents are most common?
- Which conversations fail?
- How long are users waiting?
- Where do users abandon the flow?
- Which prompts or answers need improvement?

At the same time, the dataset should avoid exposing who the users are.

## Current State

- There is no durable normalized message history yet.
- There is no analytics dataset.
- There is no data retention policy.
- There is no anonymization or redaction pipeline.
- There is no consent or notice text for analytics/training usage.

## Scope

- Define a separate analytics/training dataset.
- Create a redaction pipeline for direct identifiers and obvious personal data.
- Store analytics records separately from operational message records.
- Track consent or legal basis before using conversations for training.
- Add retention and deletion policies.
- Add a review workflow before exporting data outside the production database.
- Keep the system generic across different bot domains.

## Recommended Data Separation

Use at least two different data areas:

```text
operational_messages
analytics_messages
```

Operational messages are for product functionality:

- answering the user
- retrying failed work
- building memory
- auditing abuse
- debugging incidents

Analytics messages are for product improvement:

- intent statistics
- anonymized or redacted examples
- evaluation datasets
- future training datasets

The analytics table should never need raw Telegram identifiers.

## Suggested Analytics Fields

```text
id
source_message_id
dataset_user_id
bot_profile
message_type
language
detected_intent
redacted_text
redaction_version
contains_sensitive_data
answer_quality_label
failure_category
created_at
retention_expires_at
approved_for_training
```

`dataset_user_id` must not be the Telegram id. If user-level grouping is needed,
use a separate random id or keyed hash that is stored separately from direct
identifiers.

## Redaction Requirements

The first version should detect and redact at least:

- phone numbers
- email addresses
- URLs
- physical addresses when possible
- national ID-like values when possible
- Telegram usernames
- obvious names when possible
- account numbers or payment card-like patterns

Example:

```text
Original:
Me llamo Juan Perez y mi email es juan@example.com

Redacted:
Me llamo <person_name> y mi email es <email>
```

The redaction process must be versioned, because future versions will improve.

## Consent And Transparency

Before using user messages for training or product analytics, define the product
policy with the project owner.

The bot should eventually explain, in simple language:

- what data is stored
- why it is stored
- whether it is used to improve the bot
- how long it is kept
- how users can request deletion or opt out

For beta, do not export conversations for model training until this policy is
clear.

## Suggested MVP Policy

- Store operational messages only as required by Issue 12.
- Generate an analytics dataset only from messages that pass redaction.
- Default `approved_for_training` to false.
- Use the analytics dataset first for manual product review and evaluation, not
  external model training.
- Add an admin-only export command or script only after access control exists.
- Do not include raw Telegram ids, invite tokens, passwords, or file ids in
  exported analytics data.

## Out Of Scope

- Do not train or fine-tune a model in this issue.
- Do not build a dashboard yet.
- Do not implement complex NLP classification unless required.
- Do not send analytics data to third-party AI providers without an explicit
  product and privacy decision.
- Do not solve long-term memory here.

## Suggested Implementation Steps

1. Confirm the message persistence model from Issue 12.
2. Add a migration for an analytics dataset table.
3. Add a redaction module, for example `src/forge_bot/privacy/redaction.py`.
4. Add deterministic unit tests for each redaction pattern.
5. Generate analytics records asynchronously after operational messages are
   persisted.
6. Store the redaction version with each record.
7. Add a configurable retention period.
8. Add an admin-only export script that writes redacted JSONL or CSV.
9. Add a manual review field such as `approved_for_training`.
10. Document the policy in `README.md` or a future privacy document.

## Acceptance Criteria

- Analytics records are stored separately from operational messages.
- Analytics records do not contain raw Telegram ids.
- Redaction covers email, phone, URL, Telegram username, and common identifier
  patterns.
- The redaction pipeline is versioned.
- Training approval defaults to false.
- Exported analytics data is redacted.
- Tests cover redaction and export behavior.
- Documentation explains that pseudonymized data can still be personal data.

## Suggested Tests

- Unit test email redaction.
- Unit test phone redaction.
- Unit test URL redaction.
- Unit test Telegram username redaction.
- Unit test that raw Telegram ids are not exported.
- Unit test that `approved_for_training` defaults to false.
- Unit test analytics generation from an operational message.
- Manual test with Spanish free-text examples.

## Files To Inspect

- Message persistence code from Issue 12
- Database migrations from Issue 1
- Role/admin code from Issues 2 and 4
- Future privacy/redaction modules
- Future export scripts

## Risks

- Free-text anonymization is imperfect.
- Users may include sensitive personal data without warning.
- A small dataset can be re-identified by unique phrasing or rare details.
- If raw and redacted data are stored together without access control, the
  privacy benefit is weak.
- Using user messages for model training without a clear policy can create legal
  and trust problems.

## Dependency

Depends on Issues 12, 14, and 18. It should be completed before exporting
conversations for model training and before relying on conversation logs for
product decisions.
